### PR TITLE
Add static_map::insert_if.

### LINKED ISSUE #116: [FEA] `static_map::insert_if`
**Is your feature request related to a problem? Please describe.**
Given a set of key/value pairs, insert `pair[i]` if predicate of `stencil[i]` returns true. The API should be similar to `static_mulimap::insert_if`:
```cpp
  /**
   * @brief Inserts key/value pairs in the range `[first, last)` if `pred`
   * of the corresponding stencil returns true.
   *
   * The key/value pair `*(first + i)` is inserted if `pred( *(stencil + i) )` returns true.
   *
   * @tparam InputIt Device accessible random access input iterator where
   * `std::is_convertible<std::iterator_traits<InputIt>::value_type,
   * static_map<K, V>::value_type>` is `true`
   * @tparam StencilIt Device accessible random access iterator whose value_type is
   * convertible to Predicate's argument type
   * @tparam Predicate Unary predicate callable whose return type must be convertible to `bool` and
   * argument type is convertible from `std::iterator_traits<StencilIt>::value_type`.
   * @param first Beginning of the sequence of key/value pairs
   * @param last End of the sequence of key/value pairs
   * @param stencil Beginning of the stencil sequence
   * @param pred Predicate to test on every element in the range `[stencil, stencil +
   * std::distance(first, last))`
   * @param stream CUDA stream used for insert
   */
  template <typename InputIt, typename StencilIt, typename Predicate>
  void insert_if(
    InputIt first, InputIt last, StencilIt stencil, Predicate pred, cudaStream_t stream = 0);
```

**Additional context**
Required by cudf semi/anti join refactoring

### PR DESCRIPTION
**Pull request by vyasr**
This PR adds support for a predicated rather than unconditional insertion into a hash map. There is some code duplication introduce here, but for this PR I've prioritized getting a working implementation together to make sure that we can move forward. Once this feature is merged I plan to look into some refactorings, but I'd like to avoid conflicts with #115 while getting a working version of `insert_if` for rapidsai/cudf#9586.

Resolves #116.

### CODE REVIEW

File: include/cuco/detail/static_map.inl [hunk 1]
```diff
@@ -76,6 +76,39 @@ void static_map<Key, Value, Scope, Allocator>::insert(InputIt first,
...
+                                                         Predicate pred,
+                                                         Hash hash,
+                                                         KeyEqual key_equal)
+{
+  auto num_keys = std::distance(first, last);
+  if (num_keys == 0) { return; }
+
+  auto const block_size = 128;
+  auto const stride     = 1;
+  auto const tile_size  = 4;
```
Comment (PointKernel) [inline include/cuco/detail/static_map.inl]:
```suggestion
  auto constexpr block_size = 128;
  auto constexpr stride     = 1;
  auto constexpr tile_size  = 4;
```

File: include/cuco/detail/static_map.inl [hunk 2]
```diff
@@ -76,6 +76,39 @@ void static_map<Key, Value, Scope, Allocator>::insert(InputIt first,
...
 }
 
+template <typename Key, typename Value, cuda::thread_scope Scope, typename Allocator>
+template <typename InputIt, typename StencilIt, typename Predicate, typename Hash, typename KeyEqual>
+void static_map<Key, Value, Scope, Allocator>::insert_if(InputIt first,
+                                                         InputIt last,
+                                                         StencilIt stencil,
+                                                         Predicate pred,
+                                                         Hash hash,
+                                                         KeyEqual key_equal)
```
Comment (PointKernel) [inline include/cuco/detail/static_map.inl]:
```suggestion
                                                         KeyEqual key_equal,
                                                         cudaStream_t stream)
```
See https://github.com/NVIDIA/cuCollections/pull/113, we are adding stream to all host bulk APIs.

File: include/cuco/detail/static_map.inl [hunk 3]
```diff
@@ -76,6 +76,39 @@ void static_map<Key, Value, Scope, Allocator>::insert(InputIt first,
...
+  auto const tile_size  = 4;
+  auto const grid_size  = (tile_size * num_keys + stride * block_size - 1) / (stride * block_size);
+  auto view             = get_device_mutable_view();
+
+  // TODO: memset an atomic variable is unsafe
+  static_assert(sizeof(std::size_t) == sizeof(atomic_ctr_type));
+  CUCO_CUDA_TRY(cudaMemsetAsync(num_successes_, 0, sizeof(atomic_ctr_type)));
+  std::size_t h_num_successes;
+
+  // TODO: Should I specialize the version with a tile size?
```
Comment (PointKernel) [inline include/cuco/detail/static_map.inl]: IIRC, all host functions are using CG-based algorithms by default. So you may want to update the kernel implementation a bit and use the CG-based (tile version) device function here.

File: include/cuco/static_map.cuh [hunk 4]
```diff
@@ -221,6 +221,35 @@ class static_map {
...
+   * std::distance(first, last))`
+   * @param stream CUDA stream used for insert
+   */
+  template <typename InputIt,
+            typename StencilIt,
+            typename Predicate,
+            typename Hash     = cuco::detail::MurmurHash3_32<key_type>,
+            typename KeyEqual = thrust::equal_to<key_type>>
+  void insert_if(
+    InputIt first, InputIt last, StencilIt stencil, Predicate pred, Hash hash = Hash{}, KeyEqual key_equal = KeyEqual{});
```
Comment (PointKernel) [inline include/cuco/static_map.cuh]:
```suggestion
    InputIt first, InputIt last, StencilIt stencil, Predicate pred, Hash hash = Hash{}, KeyEqual key_equal = KeyEqual{}, cudaStream_t stream = 0);
```

Comment (PointKernel) [reply include/cuco/static_map.cuh]: Add stream support

File: include/cuco/detail/static_map.inl [hunk 5]
```diff
@@ -76,6 +76,39 @@ void static_map<Key, Value, Scope, Allocator>::insert(InputIt first,
...
+
+  auto const block_size = 128;
+  auto const stride     = 1;
+  auto const tile_size  = 4;
+  auto const grid_size  = (tile_size * num_keys + stride * block_size - 1) / (stride * block_size);
+  auto view             = get_device_mutable_view();
+
+  // TODO: memset an atomic variable is unsafe
+  static_assert(sizeof(std::size_t) == sizeof(atomic_ctr_type));
+  CUCO_CUDA_TRY(cudaMemsetAsync(num_successes_, 0, sizeof(atomic_ctr_type)));
```
Comment (PointKernel) [inline include/cuco/detail/static_map.inl]:
```suggestion
  CUCO_CUDA_TRY(cudaMemsetAsync(num_successes_, 0, sizeof(atomic_ctr_type), stream));
```

File: include/cuco/detail/static_map.inl [hunk 6]
```diff
@@ -76,6 +76,39 @@ void static_map<Key, Value, Scope, Allocator>::insert(InputIt first,
...
+  std::size_t h_num_successes;
+
+  // TODO: Should I specialize the version with a tile size?
+  detail::insert_if<block_size>
+    <<<grid_size, block_size>>>(first, first + num_keys, num_successes_, view, stencil, pred, hash, key_equal);
```
Comment (PointKernel) [inline include/cuco/detail/static_map.inl]:
```suggestion
    <<<grid_size, block_size, 0, stream>>>(first, first + num_keys, num_successes_, view, stencil, pred, hash, key_equal);
```

File: include/cuco/detail/static_map.inl [hunk 7]
```diff
@@ -76,6 +76,39 @@ void static_map<Key, Value, Scope, Allocator>::insert(InputIt first,
...
+  CUCO_CUDA_TRY(cudaMemcpyAsync(
+    &h_num_successes, num_successes_, sizeof(atomic_ctr_type), cudaMemcpyDeviceToHost));
```
Comment (PointKernel) [inline include/cuco/detail/static_map.inl]:
```suggestion
    &h_num_successes, num_successes_, sizeof(atomic_ctr_type), cudaMemcpyDeviceToHost, stream));
```

File: include/cuco/detail/static_map.inl [hunk 8]
```diff
@@ -76,6 +76,39 @@ void static_map<Key, Value, Scope, Allocator>::insert(InputIt first,
...
+  CUCO_CUDA_TRY(cudaDeviceSynchronize());
```
Comment (PointKernel) [inline include/cuco/detail/static_map.inl]:
```suggestion
  CUCO_CUDA_TRY(cudaStreamSynchronize(stream));
```

**Review by PointKernel — COMMENTED** [review]
Looks good! Just some small changes around stream support. No need to worry about https://github.com/NVIDIA/cuCollections/pull/115 right now cause it's mainly me trying things around and not sure how it goes eventually.

File: include/cuco/detail/static_map_kernels.cuh [hunk 9]
```diff
@@ -156,6 +156,60 @@ __global__ void insert(
...
+template <std::size_t block_size,
+          typename InputIt,
+          typename atomicT,
+          typename viewT,
+          typename StencilIt,
+          typename Predicate,
+          typename Hash,
+          typename KeyEqual>
+__global__ void insert_if(
+  InputIt first, InputIt last, atomicT* num_successes, viewT view, StencilIt stencil, Predicate pred, Hash hash, KeyEqual key_equal)
```
Comment (PointKernel) [inline include/cuco/detail/static_map_kernels.cuh]: Similar to the multimap implementation, the kernel API is slightly different from the host one: https://github.com/NVIDIA/cuCollections/blob/dev/include/cuco/detail/static_multimap/kernels.cuh#L91-L124. `insert_if_n` is what we want here.

**Comment by vyasr** [comment]
@PointKernel I should probably have posted some of this in my initial PR comment. There are a number of tasks that I noticed would be worth doing but I wasn't sure whether I should include in this PR. Some of them are things you commented on, so let me list here and we can scope out this PR before I make any further changes. I'm not sure that all of these are necessarily good ideas, but they are relevant so we may as well discuss them here
- I thought about adding stream support, but I wasn't sure if it would make more sense to do it in/after #113 because otherwise in the interim we'll have just one API accepting streams. What do you think?
- I considered using the `insert_if_n` approach because we need to also look up the offset in the stencil. However, I would actually like to do a broader refactoring where we implement only a single insertion kernel `insert_if` which we call from `static_map::insert` using a trivial `return true;` predicate. Assuming the compiler optimizes that out successfully (something we'd need to benchmark) it would substantially reduce code duplication.
- I didn't enable a CG API for `insert_if`, but I think we should. I would like to benchmark the performance of the current non-tiled insert implementation against the tiled implementation with a tile size of 1. Do you know if anyone has tested that? If the performance of those looks the same, I think we should be able to reorder the template arguments to give a default `tile_size` of 1 and always use that template, which would allow us to get rid of the other version.
- This change is potentially a bit more questionable, it may work but I haven't looked deeply enough to really evaluate and maybe you'll know offhand: could we unify the kernels for `static_map` and `static_multimap`, relying on the `__device__` methods of the device_view to handle the differences between the data structures? It seems to me that, in the case of `insert` for example, if we templated the kernel on the view type we could rely on `view.insert` to handle the behavior of insertion specific to the different data types, and aside from that the kernels would be identical. Thoughts?

CC @jrhemstad in case you had thoughts too. You pointed out #110 which seems to include some of these as well, so I think it's just a matter of determining what's the appropriate scope for this PR and what I should chunk out for future work.

**Comment by jrhemstad** [comment]
> I considered using the insert_if_n approach

@PointKernel 's point is that you can't rely on all iterators to have an `operator<`, so we implement the kernel as a `begin` & size algorithm where we infer the size from the provided `first/last` inputs. 

> I would like to benchmark the performance of the current non-tiled insert implementation against the tiled implementation with a tile size of 1

We have benchmarked that, and using the CG code path with a tile size of 1 is definitely slower than the purely scalar code path. There is a significant amount of extra work in the CG code path that is eliminated in the scalar path.

**Comment by PointKernel** [comment]
- https://github.com/NVIDIA/cuCollections/pull/113 focuses on existing APIs and the current PR is introducing a new API. I'm inclined to add stream argument here so it can avoid back and forth effort between those two PRs. The temporary oddness is not a big issue IMO.
- There are significant duplicates across cuco implementations (e.g. single map, multimap, etc.). I think it makes more sense to remove/optimize them in/after the "grand" refactoring when addressing https://github.com/NVIDIA/cuCollections/issues/110. A lot of things will be changed once we figure out the base open addressing data structure and for now, it's not obvious to me how the eventual APIs look like.
- As for benchmarking, it's definitely useful when evaluating CG-based algorithms against non-CG ones. I'm actually very curious to see how much improvement we can get by using CG. Based on my experience with multimap, I think the non-CG version might be more proper for single map. My big lesson here is that every parameter matters (CG vs non-CG, CG size, occupancy, data type, key distribution) so it's really easy to make misleading conclusions by overfitting one specific use case.

**Comment by vyasr** [comment]
@PointKernel I'm not sure why, but the Github UI is not permitting me to re-request review here (I tried a few different browsers). It's unclear whether that's a result of some restriction on the repo or just something wrong with Github right now. This should be ready for another pass.

**Comment by PointKernel** [comment]
> @PointKernel I'm not sure why, but the Github UI is not permitting me to re-request review here (I tried a few different browsers). It's unclear whether that's a result of some restriction on the repo or just something wrong with Github right now. This should be ready for another pass.

I guess we need to be in some specific groups cause I can't do it either. @jrhemstad maybe the time to grow cuco "maintainer" group?

File: include/cuco/detail/static_map_kernels.cuh [hunk 10]
```diff
@@ -156,6 +156,60 @@ __global__ void insert(
   if (threadIdx.x == 0) { *num_successes += block_num_successes; }
 }
 
+/**
+ * @brief Inserts all key/value pairs in the range `[first, last)`.
```
Comment (PointKernel) [inline include/cuco/detail/static_map_kernels.cuh]:
```suggestion
 * @brief Inserts key/value pairs in the range `[first, first + n)` if `pred` of the
 * corresponding stencil returns true.
```

File: include/cuco/detail/static_map_kernels.cuh [hunk 11]
```diff
@@ -156,6 +156,60 @@ __global__ void insert(
...
+ *
+ * If multiple keys in `[first, last)` compare equal, it is unspecified which
+ * element is inserted.
+ *
+ * @tparam block_size
```
Comment (PointKernel) [inline include/cuco/detail/static_map_kernels.cuh]:
```suggestion
 * @tparam block_size The size of the thread block
 * @tparam tile_size The number of threads in the Cooperative Groups used to perform insert
```

File: include/cuco/detail/static_map_kernels.cuh [hunk 12]
```diff
@@ -156,6 +156,60 @@ __global__ void insert(
...
+ * @tparam InputIt Device accessible input iterator whose `value_type` is
+ * convertible to the map's `value_type`
+ * @tparam atomicT Type of atomic storage
+ * @tparam viewT Type of device view allowing access of hash map storage
```
Comment (PointKernel) [inline include/cuco/detail/static_map_kernels.cuh:171]:
```suggestion
 * @tparam viewT Type of device view allowing access of hash map storage
 * @tparam StencilIt Device accessible random access iterator whose value_type is
 * convertible to Predicate's argument type
 * @tparam Predicate Unary predicate callable whose return type must be convertible to `bool`
 * and argument type is convertible from `std::iterator_traits<StencilIt>::value_type`
```

File: include/cuco/detail/static_map_kernels.cuh [hunk 13]
```diff
@@ -156,6 +156,60 @@ __global__ void insert(
...
+ * @tparam Hash Unary callable type
+ * @tparam KeyEqual Binary callable type
+ * @param first Beginning of the sequence of key/value pairs
+ * @param last End of the sequence of key/value pairs
```
Comment (PointKernel) [inline include/cuco/detail/static_map_kernels.cuh]:
```suggestion
 * @param n Number of elements to insert
```

File: include/cuco/detail/static_map_kernels.cuh [hunk 14]
```diff
@@ -156,6 +156,60 @@ __global__ void insert(
...
+ * @param num_successes The number of successfully inserted key/value pairs
+ * @param view Mutable device view used to access the hash map's slot storage
```
Comment (PointKernel) [inline include/cuco/detail/static_map_kernels.cuh:181]:
```suggestion
 * @param view Mutable device view used to access the hash map's slot storage
 * @param stencil Beginning of the stencil sequence
 * @param pred Predicate to test on every element in the range `[s, s + n)`
```

File: include/cuco/detail/static_map_kernels.cuh [hunk 15]
```diff
@@ -156,6 +156,60 @@ __global__ void insert(
...
+  auto tile      = cg::tiled_partition<tile_size>(cg::this_thread_block());
+  auto tid = block_size * blockIdx.x + threadIdx.x;
+  auto i = tid / tile_size;
+
+  while (i < n) {
+    if (pred(*(stencil + i))) {
+      typename viewT::value_type const insert_pair{*(first + i)};
+      if (view.insert(tile, insert_pair, hash, key_equal)) { thread_num_successes++; }
+    }
+    i += gridDim.x * block_size;
```
Comment (PointKernel) [inline include/cuco/detail/static_map_kernels.cuh]:
```suggestion
    i += (gridDim.x * block_size) / tile_size;
```

The CG-based algorithm uses `tile_size` threads to handle each input pair (one tile for one pair insertion) so the actual grid stride is `girdDim x blockDim / tileSize`

File: include/cuco/detail/static_map_kernels.cuh [hunk 16]
```diff
@@ -156,6 +156,60 @@ __global__ void insert(
...
+__global__ void insert_if_n(
+  InputIt first, std::size_t n, atomicT* num_successes, viewT view, StencilIt stencil, Predicate pred, Hash hash, KeyEqual key_equal)
+{
+  typedef cub::BlockReduce<std::size_t, block_size> BlockReduce;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+  std::size_t thread_num_successes = 0;
+
+  auto tile      = cg::tiled_partition<tile_size>(cg::this_thread_block());
+  auto tid = block_size * blockIdx.x + threadIdx.x;
+  auto i = tid / tile_size;
```
Comment (PointKernel) [inline include/cuco/detail/static_map_kernels.cuh]: Looks like code formatting is missing

File: include/cuco/detail/static_map_kernels.cuh [hunk 17]
```diff
@@ -156,6 +156,60 @@ __global__ void insert(
...
+      typename viewT::value_type const insert_pair{*(first + i)};
+      if (view.insert(tile, insert_pair, hash, key_equal)) { thread_num_successes++; }
+    }
+    i += gridDim.x * block_size;
+  }
+
+  // compute number of successfully inserted elements for each block
+  // and atomically add to the grand total
+  std::size_t block_num_successes = BlockReduce(temp_storage).Sum(thread_num_successes);
+  if (threadIdx.x == 0) { *num_successes += block_num_successes; }
```
Comment (PointKernel) [inline include/cuco/detail/static_map_kernels.cuh]:
```suggestion
  if (threadIdx.x == 0) {
    num_matches->fetch_add(block_num_matches, cuda::std::memory_order_relaxed);
  }
```
Not necessarily a change request: `+=` will work, using `fetch_add` with relaxed memory order may bring minor but visible speedups.

File: include/cuco/static_map.cuh [hunk 18]
```diff
@@ -221,6 +221,35 @@ class static_map {
...
+   * convertible to Predicate's argument type
+   * @tparam Predicate Unary predicate callable whose return type must be convertible to `bool` and
+   * argument type is convertible from `std::iterator_traits<StencilIt>::value_type`.
+   * @tparam Hash Unary callable type
+   * @tparam KeyEqual Binary callable type
+   * @param first Beginning of the sequence of key/value pairs
+   * @param last End of the sequence of key/value pairs
+   * @param stencil Beginning of the stencil sequence
+   * @param pred Predicate to test on every element in the range `[stencil, stencil +
+   * std::distance(first, last))`
```
Comment (PointKernel) [inline include/cuco/static_map.cuh:242]:
```suggestion
   * std::distance(first, last))`
   * @param hash The unary function to hash each key
   * @param key_equal The binary function to compare two keys for equality
```

File: include/cuco/detail/static_map_kernels.cuh (hunk 15 above)
Comment (vyasr) [reply include/cuco/detail/static_map_kernels.cuh]: Good catch, thanks. I completely missed that this was how the algorithm was handling the insertion.

File: include/cuco/detail/static_map_kernels.cuh (hunk 16 above)
Comment (vyasr) [reply include/cuco/detail/static_map_kernels.cuh]: We should add a style check to CI. I'll make an issue.

**Comment by vyasr** [comment]
@PointKernel Thanks for those comments on CG and atomics, I haven't used either yet and I learned a few new things. Should be helpful in future refactorings/optimizations.

**Review by PointKernel — APPROVED** [review]

**Review by jrhemstad — APPROVED** [review]

**Pull request merged**